libcalico-go/lib/backend/syncersv1/dedupebuffer/dedupe_buffer_test.go

```diff
@@ -0,0 +1,250 @@
+			Expect(rec.UpdatesSeen()).To(Equal([]string{
+				// WaitForDatastore gets skipped since it's in the same batch.
+				string(api.ResyncInProgress),
+				"foo=bar",
+				"foo2=bar2",
+				string(api.InSync),
+			}))
+
+			// Now send in a deletion.
+			rec.ResetUpdatesSeen()
+			onUpdates([]api.Update{KVUpdate("foo", "")})
+			applyNextBatchSync(d, rec)
+			Expect(rec.FinalValues()).To(Equal(map[string]string{
+				"foo2": "bar2",
+			}))
+			Expect(rec.FinalSyncState()).To(Equal(api.InSync))
+			Expect(rec.UpdatesSeen()).To(Equal([]string{
+				"foo=",
+			}))
+
+			// Now send in a deletion, which is reverted before being sent.
+			rec.ResetUpdatesSeen()
+			onUpdates([]api.Update{KVUpdate("foo2", "")})
+			onUpdates([]api.Update{KVUpdate("foo2", "bar3")})
+			applyNextBatchSync(d, rec)
+			Expect(rec.FinalValues()).To(Equal(map[string]string{
+				"foo2": "bar3",
+			}))
+			Expect(rec.FinalSyncState()).To(Equal(api.InSync))
+			Expect(rec.UpdatesSeen()).To(Equal([]string{
+				"foo2=bar3",
+			}))
+
+			// Update both keys, should work as normal.
+			rec.ResetUpdatesSeen()
+			onUpdates([]api.Update{KVUpdate("foo", "bar")})
+			onUpdates([]api.Update{KVUpdate("foo2", "bar2")})
+			applyNextBatchSync(d, rec)
+			Expect(rec.FinalValues()).To(Equal(map[string]string{
+				"foo":  "bar",
+				"foo2": "bar2",
+			}))
+			Expect(rec.FinalSyncState()).To(Equal(api.InSync))
+			Expect(rec.UpdatesSeen()).To(Equal([]string{
+				// WaitForDatastore gets skipped since it's in the same batch.
+				"foo=bar",
+				"foo2=bar2",
+			}))
+		})
+	}
+}
+
+func TestDedupeBuffer_SyncWithDupes(t *testing.T) {
+	for _, onUpdatesVersion := range []string{"KeysKnown", "KeysNotKnown"} {
+		t.Run(onUpdatesVersion, func(t *testing.T) {
+			RegisterTestingT(t)
+			d := New()
+			onUpdates := wrapOnUpdates(d, onUpdatesVersion)
+			rec := NewReceiver()
+
+			d.OnStatusUpdated(api.WaitForDatastore)
+			d.OnStatusUpdated(api.ResyncInProgress)
+			onUpdates([]api.Update{KVUpdate("foo", "bar")})
+			onUpdates([]api.Update{KVUpdate("foo2", "bar2")})
+			onUpdates([]api.Update{KVUpdate("foo3", "bar3")})
+			onUpdates([]api.Update{KVUpdate("foo", "bar3")})
+			onUpdates([]api.Update{KVUpdate("foo2", "")})
+			d.OnStatusUpdated(api.InSync)
+			d.OnStatusUpdated(api.ResyncInProgress)
+			d.OnStatusUpdated(api.ResyncInProgress)
+			d.OnStatusUpdated(api.InSync)
+
+			applyNextBatchSync(d, rec)
+
+			Expect(rec.FinalValues()).To(Equal(map[string]string{
+				"foo":  "bar3",
+				"foo3": "bar3",
+			}))
+			Expect(rec.FinalSyncState()).To(Equal(api.InSync))
+			Expect(rec.UpdatesSeen()).To(Equal([]string{
+				// WaitForDatastore skipped.
+				string(api.ResyncInProgress),
+				"foo=bar3", // Update leap-frogs the original value.
+				"foo3=bar3",
+				string(api.InSync),
+			}))
+		})
+	}
+}
+
+func applyNextBatchSync(d *DedupeBuffer, r *Receiver) {
+	Expect(d.pendingUpdates.Len()).NotTo(BeZero(), "Nothing on queue, sendNextBatchToSink would block")
+	d.sendNextBatchToSink(r)
```

Comment:
Oh, maybe this is why this code wasn't inlined...

Reply:
Yep, I had it inlined and broke it out for ease of testing.